flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java

```diff
@@ -181,11 +181,13 @@ public boolean isImmutableType() {
 			}
 		}
 
-		if (stateful) {
-			return new PojoSerializer<T>(clazz, duplicateFieldSerializers, fields, executionConfig);
-		} else {
-			return this;
+		if (!stateful) {
+			// as a small memory optimization, we can share the same object between instances
+			duplicateFieldSerializers = fieldSerializers;
 		}
+
+		// we must create a new instance, otherwise the subclassSerializerCache can create concurrency problems
```

Comment:
Nope, I don't understand this change :/

The stacktrace shows that the KryoSerializer is used by multiple threads concurrently. `KryoSerializer#duplicate` always returns a new instance, so this method should've already returned a new instance since `stateful` is true.

Reply:
The point is that the Kryo serializer is not coming at all from the `fieldSerializers` that we duplicate. It comes from `subclassSerializerCache` and nothing is duplicated from there, which it also should not.